src/backend/storage/buffer/bufmgr.c

```diff
@@ -1531,6 +1531,8 @@ MarkBufferDirty(Buffer buffer)
 	if (!(old_buf_state & BM_DIRTY))
 	{
 		VacuumPageDirty++;
+		//Zenith XXX: Why do we clear it only if previous state was not dirty?
```

Comment:
Hints are just optimization, so it is not a problem to loose hints update, isn't it?

Reply:
> Hints are just optimization, so it is not a problem to loose hints update, isn't it?

Sure, MarkBufferDirtyHint() only used for optimizations that we can afford to lose. 
I was thinking about two things. First is the PD_ALL_VISIBLE hint bit - vm functions expect it to be set whenever corresponding vm bit is set, so I think we'd better not skip it.
And second is the performance impact - since hint changes are not wal-logged, we effectively never persist them and lose them every time when the page is evicted. Though, I agree that maybe it's too early to think about such optimizations.


And what about the first question? Why do we clear PD_WAL_LOGGED only if the page wasn't dirty before?
I cannot convince myself that it is correct.
Consider the following scenario:
1. change page content, markBufferDirty() - set BM_DIRTY, clean PD_WAL_LOGGED
2. log --> XLogRegisterBlock - set PD_WAL_LOGGED
3. change page content, markBufferDirty() again - buf is already dirty, this call won't clean PD_WAL_LOGGED
4. don't wal-log page (i.e. because of create index tricks)
5. evict page. zenith_wallog_page() won't issue a log record, since PD_WAL_LOGGED is set. And we lost our second change of the page.

I am not sure that such a scenario really exists in the current codebase, but I don't think we can prove otherwise.

Reply:
I think you are tight: we should always clear PD_WAL_LOGGED in MarkBufferDirty